Lib/asyncio/tasks.py

```diff
@@ -438,9 +438,16 @@ async def wait_for(fut, timeout):
 
     If the wait is cancelled, the task is also cancelled.
 
+    If the task supresses the cancellation and returns a value instead,
+    that value is returned.
+
     This function is a coroutine.
     """
 
+    # When timeout <= 0, `asyncio.timeout` raises `TimeoutError`
+    # whereas `wait_for` disables the timeout so we special case
+    # it here to preserve the behavior.
```

Comment:
The way I read the code following, it is not "disabling" the timeout, rather it is cancelling the future (`_cancel_and_wait`). I know there is a difference between this and what timeout() does, but I still don't understand the difference. :-( Can you explain it to me?

Reply:
That comment is wrong, sorry for that, removed now.

This is the simplest case I can think of which explains the behaviour:

```py
import asyncio


async def test_timeout():
    func_started = False

    async def func():
        nonlocal func_started
        func_started = True

    async with asyncio.timeout(0):
        await func()

    assert func_started

asyncio.run(test_timeout())

async def test_waitfor():
    func_started = False

    async def func():
        nonlocal func_started
        func_started = True

    try:
        await asyncio.wait_for(func(), 0)
    except asyncio.TimeoutError:
        assert not func_started
    else:
        assert False


asyncio.run(test_waitfor())
```

On main both asserts pass.

On this branch if I remove the `timeout <= 0` branch then the assert in `test_waitfor` fails.

Reply:
I have added `test_waitfor` code as comment as why the special case is needed.